ecommerce/api_test.py

```diff
@@ -719,3 +734,92 @@ def test_not_redeemable(self):
             assert pick_coupons(self.user) == []
         for coupon in Coupon.objects.all().exclude(id=self.not_auto_or_attached_coupon.id):
             _is_coupon_redeemable.assert_any_call(coupon, self.user)
+
+
+class PriceTests(MockedESTestCase):
+    """
+    Tests for calculating prices
+    """
+
+    def test_calculate_run_price_no_coupons(self):
+        """
+        If there are no coupons for this program the price should be what get_formatted_course_price returned
+        """
+        course_run, user = create_purchasable_course_run()
+        # This coupon is for a different program
+        coupon = CouponFactory.create()
+        UserCoupon.objects.create(coupon=coupon, user=user)
+        discounted_price = 5
+        fa_price = get_formatted_course_price(course_run.course.program.programenrollment_set.first())['price']
+        with patch('ecommerce.api.calculate_coupon_price', autospec=True) as _calculate_coupon_price:
+            _calculate_coupon_price.return_value = discounted_price
+            assert calculate_run_price(course_run, user) == (fa_price, None)
+        assert _calculate_coupon_price.called is False
+
+    def test_no_program_enrollment(self):
+        """
+        If a user is not enrolled a 404 should be raised when getting the price
+        """
+        course_run, user = create_purchasable_course_run()
+        ProgramEnrollment.objects.filter(program=course_run.course.program, user=user).delete()
+        try:
+            calculate_run_price(course_run, user)
+        except Http404:
+            pass
+
+    def test_calculate_run_price_coupon(self):
+        """
+        If there is a coupon calculate_run_price should use calculate_coupon_price to get the discounted price
+        """
+        course_run, user = create_purchasable_course_run()
+        coupon = CouponFactory.create(content_object=course_run)
+        UserCoupon.objects.create(coupon=coupon, user=user)
+        discounted_price = 5
+        with patch('ecommerce.api.calculate_coupon_price', autospec=True) as _calculate_coupon_price:
+            _calculate_coupon_price.return_value = discounted_price
+            assert calculate_run_price(course_run, user) == (discounted_price, coupon)
+        fa_price = get_formatted_course_price(course_run.course.program.programenrollment_set.first())['price']
+        _calculate_coupon_price.assert_called_with(coupon, fa_price, course_run.edx_course_key)
```

Comment:
i think it would be better just to create an actual coupon with a fixed discount instead of patching `calculate_coupon_price`. seems pretty straightforward. either way, i think we should get rid of this `assert_called_with`. we've already tested that the function is invoked and returns values we expect. all i see this doing is making it more difficult for us to change implementation